src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalSessionSocket.java

```diff
@@ -309,15 +326,42 @@ public void dispatchInput(int inputSequence,
       default:
          break;
       }
-      
    }
    
    /**
     * Send output to the terminal emulator.
     * @param output text to send to the terminal
+    * @param detectLocalEcho local-echo detection
     */
-   public void dispatchOutput(String output)
+   public void dispatchOutput(String output, boolean detectLocalEcho)
    {
+      if (detectLocalEcho && !localEcho_.isEmpty())
+      {
+         String lastOutput = "";
+         while (!localEcho_.isEmpty() && lastOutput.length() < output.length())
```

Comment:
I wasn't aware of the set -o vi case; definitely busted for that and to support it I think I'd have to specifically track the bash "vi" state between input mode and command mode in the client. Need to put knowledge of that state machine and what key sequences change it into the client code.

Also discovered issues with default emacs mode; can move the cursor around via the Ctrl+? sequences, but as soon as you try to insert anything it gets confused because of the added escape sequences around the characters that have already been local-echoed. So, to pull this off for Emacs, need to do client-side parsing of the escape sequences, or maybe be able to tell if cursor has moved from EOL and stop doing local-echo.